### PR TITLE
Add GNMT Config to Autopilot

### PR DESCRIPTION
**Pull request by xinyu-intel**
Hi, we trained a de-en gnmt model with 8 layers LSTM encoder & decoder using sockeye. The validation bleu on `newstests2016.de` has reached up to 26.91 both on CPU and GPU. Also, the translate performance is very good on CPU (around 24 sentences/sec on Skylake 8180 one socket) after some optimizations. We are going to add the model config to autopilot to enable out-of-box GNMT task.
@pengzhao-intel @TaoLv
#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [x] Were system tests modified? If so did you run these at least 5 times to account for the variation across runs?
- [x] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by xinyu-intel** [comment]
The model in this PR is aligned with #479

File: contrib/autopilot/models.py [hunk 1]
```diff
@@ -14,9 +14,15 @@
 # Names model types
 MODEL_NONE = "none"
 MODEL_TRANSFORMER = "transformer"
+MODEL_GNMT = "gnmt"
+
+# Names model test settings
+MODEL_TEST_ARGS_TRANSFORMER = "ttransformer"
+MODEL_TEST_ARGS_GNMT = "tgnmt"
```
Comment (mjdenkowski) [inline contrib/autopilot/models.py]: We shouldn't need `MODEL_TEST_ARGS_TRANSFORMER ` or `MODEL_TEST_ARGS_GNMT` if we already have `MODEL_TRANSFORMER` and `MODEL_GNMT`.

File: contrib/autopilot/models.py [hunk 2]
```diff
@@ -14,9 +14,15 @@
...
 
 # Named decoding settings
 DECODE_STANDARD = "standard"
+DECODE_GNMT = "dgnmt"
```
Comment (mjdenkowski) [inline contrib/autopilot/models.py]: `dgnmt` -> `gnmt`

File: contrib/autopilot/models.py [hunk 3]
```diff
@@ -59,22 +65,62 @@
...
-    "--transformer-feed-forward-num-hidden=16",
-    "--num-embed=16:16",
-    "--num-words=16:16",
-    "--batch-type=sentence",
-    "--batch-size=1",
-    "--max-updates=4",
-    "--checkpoint-frequency=2",
-]
+MODEL_TEST_ARGS = {
+    MODEL_TEST_ARGS_TRANSFORMER: [
```
Comment (mjdenkowski) [inline contrib/autopilot/models.py]: `MODEL_TEST_ARGS_TRANSFORMER` -> `MODEL_TRANSFORMER`

File: contrib/autopilot/models.py [hunk 4]
```diff
@@ -59,22 +65,62 @@
...
+        "--transformer-feed-forward-num-hidden=16",
+        "--num-embed=16:16",
+        "--num-words=16:16",
+        "--batch-type=sentence",
+        "--batch-size=1",
+        "--max-updates=4",
+        "--checkpoint-frequency=2",
+    ],
+
+    MODEL_TEST_ARGS_GNMT: [
```
Comment (mjdenkowski) [inline contrib/autopilot/models.py]: `MODEL_TEST_ARGS_GNMT` -> `MODEL_GNMT`

File: contrib/autopilot/autopilot.py [hunk 5]
```diff
@@ -43,8 +43,9 @@
 from contrib.autopilot.tasks import TEXT_UTF8_TOKENIZED
 from contrib.autopilot.tasks import RAW_FILES
 from contrib.autopilot.tasks import Task, TASKS
-from contrib.autopilot.models import MODELS, MODEL_NONE, MODEL_TEST_ARGS
-from contrib.autopilot.models import DECODE_ARGS, DECODE_STANDARD
+from contrib.autopilot.models import MODELS, MODEL_NONE
+from contrib.autopilot.models import MODEL_TEST_ARGS, MODEL_TEST_ARGS_TRANSFORMER, MODEL_TEST_ARGS_GNMT
```
Comment (mjdenkowski) [inline contrib/autopilot/autopilot.py]: We shouldn't need `MODEL_TEST_ARGS_TRANSFORMER` or `MODEL_TEST_ARGS_GNMT`.

File: contrib/autopilot/autopilot.py [hunk 6]
```diff
@@ -863,6 +864,8 @@ def main():
                             help="Pre-defined data set for model training.")
     arg_parser.add_argument("--model", type=str, choices=sorted(MODELS.keys()),
                             help="Type of translation model to train.")
+    arg_parser.add_argument("--model-test-settings", type=str, choices=sorted(MODEL_TEST_ARGS.keys()), default=MODEL_TEST_ARGS_TRANSFORMER,
+                            help="Model test settings. Default: %(default)s.")
```
Comment (mjdenkowski) [inline contrib/autopilot/autopilot.py]: We shouldn't need to specify this as we can use the model type as the index into `MODEL_TEST_ARGS`.

**Comment by fhieber** [comment]
I am not sure we should call this model 'gnmt'. I assume you are referring to the models in https://arxiv.org/pdf/1609.08144.pdf. Do they also use an embedding size of 256 for example?
I would prefer renaming this configuration to something like '8-layer rnn' to highlight the different architecture to the transformer. We can add a comment or description that this configuration resembles the 'GNMT' architecture described in https://arxiv.org/pdf/1609.08144.pdf.

**Comment by xinyu-intel** [comment]
@fhieber Thanks for advice. I think this config is not exactly the same as GNMT described in paper but is very similar with it and has very good performance and accuracy. However, GNMT in paper also has more than one combinations. It's main idea is to use LSTMs as encoder and decoder and  I think it is free for users to choose 256 or 512 as embedding size and any other configs like encoder/decoder layer numbers, hidden layer numbers, right?

**Comment by mjdenkowski** [comment]
Do the settings not exactly match the GNMT paper?  I like the idea of having presets that correspond to particular published models, but I also agree that we shouldn't call something "GNMT" when it's not.  Perhaps something like `gnmt_like` to indicate this is as close as we can get in Sockeye and see comparable BLEU scores?

**Comment by xinyu-intel** [comment]
@mjdenkowski Thanks for your advice, I've changed the mdoel name to `gnmt_like` and add a comment when using this config `This is an 8 layer LSTMs model which can resembles the 'GNMT' architecture.` And all tests have passed:)

**Comment by TaoLv** [comment]
I think implementing more famous published models can really help improve the community of framework and toolkits like sockeye. GNMT is one of this kind of famous models. So we are trying to implement and train it on the top of mxnet and sockeye. I know GluonNLP is also trying to do something similar.
Here in this PR, we want to present a GNMT model as close as possible to the original paper, and also want to make it comparable to the TF implementation for both accuracy and performance. As @xinyu-intel mentioned, we have got 26.91 for BLEU on both CPU and GPU. If you are interested in the performance of this model, I think @xinyu-intel can also share some initial data for it. Anyway, I think it would be great if we can figure out that mxnet and sockeye have the same GNMT model which has better performance compared with other frameworks, right?
So if you find anything in this PR does not align with the original GNMT paper or TF implementation, please kindly point it out. Or if you have a more general implementation for GNMT, please kindly share it with us if possible. We are pleased and looking forward to working with the sockeye community to make this happen.

**Comment by fhieber** [comment]
Totally agree with you @TaoLv and I think its great to have a well-working RNN-type model readily available in Sockeye for others to use! My point was to ask how sure you (or others) are that this configuration is as close as possible to the GNMT model (given that there was a discussion around the correct parameters in #479). I haven't looked into replicating GNMT models in Sockeye, but did for the transformer and there, many details (down to embedding initialization, scaling etc.) are crucial to reproduce published results.
If the config in this PR does equally or better than published results in the GNMT paper, this is awesome! But we should try to be extremely clear on what number in the paper we are comparing against, how we compute it, how - to the best of our knowledge - the setup differs from the published paper, and what language pair/dataset we are reporting this on.
Again, more presets/configs for Sockeye autopilot are super welcome and helpful, but in my experience replicating published models across different toolkits and frameworks in a fair way can be quite complex, and before we make any claim that GNMT models in MXNet/Sockeye do better than previously published results, we need to be absolutely sure about it.

That said, calling the current config `gnmt_like` is absolutely fine with me and you can always iterate on it to make it more precise (maybe to eventually call it `gnmt`).

**Comment by TaoLv** [comment]
You're right @fhieber. It's not an easy task to implement a publised model and reproduce the accuracy of it. That's why we want to provide model zoo, pre-trained models and fixed configurations to end users. I'm quite OK if you want to merge this PR as a `gnmt_like` configuration for now. And I still think it's a good start point for us to push this effort forward to create a more general, comparable and efficient GNMT model based mxnet and sockeye.

File: contrib/autopilot/autopilot.py [hunk 7]
```diff
@@ -721,6 +721,8 @@ def report_data(file_sets):
 
     logging.info("=== Train translation model ===")
     logging.info("Model: %s", args.model)
+    if args.model == "gnmt_like":
```
Comment (mjdenkowski) [inline contrib/autopilot/autopilot.py]: `"gnmt_like"` -> `MODEL_GNMT`

File: contrib/autopilot/autopilot.py [hunk 8]
```diff
@@ -721,6 +721,8 @@ def report_data(file_sets):
...
+        logging.info("WARNING: This is an 8 layer LSTMs model which can resembles the 'GNMT' architecture.")
```
Comment (mjdenkowski) [inline contrib/autopilot/autopilot.py]: This could be worded: `"NOTE: This is an 8 layer LSTM model similar (but not exactly identical) to the 'GNMT' architecture."`

Comment (xinyu-intel) [reply contrib/autopilot/autopilot.py]: @mjdenkowski Thanks, changed:)

**Comment by mjdenkowski** [comment]
Can you update the minor version and changelog?

**Comment by xinyu-intel** [comment]
@mjdenkowski added

**Comment by TaoLv** [comment]
Glad to see this PR was merged. Now, sockeye autopilot supports both `transformer` and `gnmt_like`. So do you think we need to adjust the README of autopilot to promote that and tell users how to run `gnmt_like` and reproduce the accuracy. @xinyu-intel @fhieber
